text/0000-unicode-idents.md

```diff
@@ -0,0 +1,161 @@
+
+Private projects can use any script and language the developer(s) desire. It is still a good idea (as with any language feature) not to overdo it.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Identifiers in Rust are based on the [Unicode® Standard Annex #31 Unicode Identifier and Pattern Syntax][TR31]. Rust compilers shall use at least Revision 27 of the standard.
+
+The lexer defines identifiers as:
+
+> **<sup>Lexer:<sup>**  
+> IDENTIFIER_OR_KEYWORD:  
+> &nbsp;&nbsp; XID_Start&nbsp;XID_Continue<sup>\*</sup>  
+> &nbsp;&nbsp; | `_` XID_Continue<sup>+</sup>  
+>  
+> IDENTIFIER :  
+> IDENTIFIER_OR_KEYWORD <sub>*Except a [strict] or [reserved] keyword*</sub>
+
+`XID_Start` and `XID_Continue` are used as defined in the aforementioned standard. The definition of identifiers is forward compatible with each successive release of Unicode as only appropriate new characters are added to the classes but none are removed.
+
+Two identifiers X, Y are considered to be equal if their [NFKC forms][TR15] are equal: NFKC(X) = NFKC(Y).
+
+A `unicode_idents` lint is added to the compiler. This lint is `allow` by default. The lint checks if any identifier in the current context contains a codepoint with a value equal to or greater than 0x80 (outside ASCII range). Not only locally defined identifiers are checked but also those imported from other crates and modules into the current context. 
+
+## Confusable detection
+
+Rust compilers should detect confusingly similar Unicode identifiers and warn the user about it.
+
+Note: This is *not* a mandatory for all Rust compilers as it requires considerable implementation effort and is not related to the core function of the compiler. It rather is a tool to detect accidental misspellings and intentional homograph attacks.
+
+A new `confusable_unicode_idents` lint is added to the compiler. The default setting is `warn`.
+
+Note: The confusable detection is set to `warn` instead of `deny` to enable forward compatibility. The list of confusable characters will be extended in the future and programs that were once valid would fail to compile.
+
+The confusable detection algorithm is based on [Unicode® Technical Standard #39 Unicode Security Mechanisms Section 4 Confusable Detection][TR39Confusable]. For every distinct identifier X in the current scope execute the function `skeleton(X)`. If there exist two distinct identifiers X and Yin the same crate where `skeleton(X) = skeleton(Y)` report it.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* "ASCII is enough for anyone." As source code should be written in English and in English only (source: various people) no charactes outside the ASCII range are needed to express identifiers. Therefore support for Unicode identifiers introduces unnecceray complexity to the compiler.
```

Comment:
"source code should be written in English and in English only" is not an argument, it's an opinion - one that is easily shown to be partial by pointing out how ballistic people can go when e.g. variable names don't convey the correct meaning of the variable; now think if, because you're not a good English speaker, they convey *no* meaning to you at all...

Reply:
@kimhyunkang exposes even more reasons why this stance is terrible for diversity/inclusion in https://github.com/rust-lang/rfcs/pull/2457#issuecomment-395148135

Reply:
Also keep in mind that especially for younger programmers (whose native language is not English) this doesn't hold. People try out programming while they are in school and have courses on English but aren't nearly fluent yet. It's much easier to write code in your native language at that point in life, so you can focus on learning how to code instead of also needing to learn English at the same time. 

Source: I'm Dutch and although all code I write now is English I have definitely written Dutch code in my younger years when I wasn't fluent enough in English yet and was learning how to code. Luckily I wasn't held back by non ASCII identifiers, because

Reply:
> "source code should be written in English and in English only" is not an argument, it's an opinion

That's true. But I still still hear this argument often. The underlying argument is that one needs to write in English so other people can read it. And even if you write code just for yourself you should write in English to learn it.

If you think this statement should be rephrased please suggest something else.